kafka-cron/producer/main.go

```diff
@@ -0,0 +1,192 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+
+	"github.com/Shopify/sarama"
+	"github.com/berkeli/kafka-cron/types"
+	"github.com/goccy/go-yaml"
+	"github.com/google/uuid"
+	"github.com/robfig/cron/v3"
+)
+
+var (
+	brokerList  = kingpin.Flag("brokerList", "List of brokers to connect").Default("localhost:9092").Strings()
+	topicPrefix = kingpin.Flag("topicPrefix", "Topic prefix, e.g. jobs will create topics for each cluster in the format jobs-cluster-a").Default("jobs").String()
+	cronPath    = kingpin.Flag("cronPath", "Path to cron file").Default("./cron.yaml").String()
+	partitions  = kingpin.Flag("partitions", "Number of partitions").Default("1").Int32()
+)
+
+func main() {
+	InitMonitoring(2112)
+
+	kingpin.Parse()
+	config := sarama.NewConfig()
+
+	admin, err := sarama.NewClusterAdmin(*brokerList, config)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	config.Producer.Return.Successes = true
+	producer, err := sarama.NewSyncProducer(*brokerList, config)
+	if err != nil {
+		log.Panic(err)
+	}
+	defer func() {
+		if err := producer.Close(); err != nil {
+			log.Panic(err)
+		}
+		if err := admin.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+	cmds, err := ReadConfig(*cronPath)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	err = CreateTopics(admin, cmds)
```

Comment:
What are the trade-offs of having your producer create the topics, vs having your producer assume the topics are already created (and error if not)?

Reply:
Excellent question although TBF I did suggest they do this, mainly to simplify life with respect to not having to have a producer topic setup script of some kind to get started.
```
We will also need to [create a Kafka topic](https://kafka.apache.org/documentation/#quickstart_createtopic). In a production environment we would probably use separate tooling to manage topics (perhaps Terraform), but for this project, we can [create our Kafka topic using code](https://github.com/confluentinc/examples/blob/7.3.0-post/clients/cloud/go/producer.go#L39).
```